src/drug_discovery/structures/protein.py

```diff
@@ -730,6 +634,51 @@ def extract_metals_and_cofactors(self) -> Tuple[list[str], list[str]]:
 
         return metal_resnames, cofactor_resnames
 
+    def extract_ligand(self, exclude_resnames: Optional[set] = None):
+        """
+        Extracts ligand(s) from a PDB file using RDKit.
+
+        Args:
+            exclude_resnames (set): Residue names to exclude (e.g., water).
+
+        """
+
+        from rdkit import Chem
+
+        if exclude_resnames is None:
+            exclude_resnames = {"HOH"}
+
+        ligand_lines = []
+        with open(self.file_path, "r") as f:
+            for line in f:
+                if line.startswith("HETATM"):
```

Comment:
Preserve bonding. If you copy only HETATM lines you lose CONECT records; RDKit may guess bonds poorly. Include matching CONECTs.

Also: consider skipping non‑primary altLocs and filtering tiny ions.

```python
if line.startswith("HETATM"):
    resname = line[17:20].strip()
    altloc = line[16].strip()

    if resname in exclude_resnames:
        continue

    if altloc not in ("", "A"):  # skip altLocs other than primary
        continue

    ligand_lines.append(line)
```

Reply:
thanks, fixed